On going improvements and tweaks from hardware testing.

christhechris opened the pull request:
- update wavelength calibration to use normal crop, better fit visualisations, and logic to make sure line matches are close.
- 'update_intsphere_cube' records real exposure time of camera (not always same as what was requested, small difference due hardware).
-  enter and exit added to CameraProps to enable 'with'
- logic in tfm_setup to select appropriate rad_ref if a luminance is not set in settings file.

YiweiMao commented on openhsi/calibrate.py:
```diff
@@ -164,22 +173,41 @@ def fit_HgAr_lines(self, top_k:int = 10,
+            print(diff)
+            if diff < max_match_error:  # nm
+                closest_HgAr_line = np.append(
+                    closest_HgAr_line, HgAr_lines[np.argmin(np.abs(HgAr_lines - λ))]
```
This could made neater using standard Python lists. I'm not so comfortable with how long this function is getting and it's taking me longer to understand what's going on.  
There's lots of diagnostics printing. enable using `verbose=True`? See my suggestion below
```python
closest_HgAr_line, matching_centroid, matching_A  = [], [], []
diffs = [np.abs(HgAr_lines-λ) for λ in predicted_λ]
for i in range(len(diffs)):
    if verbose: print(f"difference HgAr_lines - λ = {diffs[i]}")
    if diffs[i] < max_match_error: # nm
        closest_HgAr_line.append( HgAr_lines[np.argmin(np.abs(HgAr_lines - predicted_λ[i]))] )
        matching_centroid.append( mu[i] ); matching_A.append( A[i] )
```

christhechris replied:
happy with something like that, I was not very happy with original version, just needed something working at the time.

christhechris replied:
<img width="758" alt="image" src="https://user-images.githubusercontent.com/782659/150454502-eea7d85f-f770-40a9-a578-6b6fd70e3d13.png">

this does not seem to work as is.

christhechris replied:
put I'm not sure my port from my code to current master was correct, let me check that.

YiweiMao replied:
I forgot the `min` in 
`diffs = [np.abs(HgAr_lines-λ) for λ in predicted_λ]`
should be
`diffs = [np.min(np.abs(HgAr_lines-λ)) for λ in predicted_λ]`
?

christhechris replied:
ahh, yup, that helps.

another issue is the indexing of lists with Numpy array 'top_A_idx'

could fix with np.asarray where matching_centroid and closest_HgAr_line are used, but it is a bit messy.

`final_fit = np.poly1d( np.polyfit(np.asarray(matching_centroid)[top_A_idx],np.asarray(closest_HgAr_line)[top_A_idx] ,3) )
 `

otherwise convert them to ndarray after building

`closest_HgAr_line= np.asarray(closest_HgAr_line)`
etc

or init as ndarray like my original one:
`closest_HgAr_line,matching_centroid,matching_A = [np.empty(0) for i in range(3)]`

christhechris replied:
fixed by converting to ndarray at end.

YiweiMao commented on openhsi/calibrate.py:
```diff
@@ -153,9 +159,12 @@ def fit_HgAr_lines(self, top_k:int = 10,
         plt.show()
 
-        # interpolate with top 3 spectral lines
+        # interactivly confirm peak wavelenght
```
wavelengths

christhechris replied:
fixed in current version

YiweiMao commented:
Using .crop to replace the old slice method is a good idea. I find some of the logic formatted with Black is very lengthy - not that it matters all that much.

christhechris commented:
on formatting, I didn't black everything, which bit were you thinking of here?

christhechris commented:
Updated and added docstring in the calibrate notebook.

on a related note, I think we should add a way to generate flat and arc image (perhaps using real sample data) in the SimulatedCamera.  With this we can then get some tests and examples running in calibrate notebook without need for hardware etc. started a issue/feature request for this #20.

YiweiMao reviewed (approved):
I'm happy with the changes including the Black formatting. Just some of the chained function calls created multiline closing ")"s which is probably inspired from C++'s closing braces. I try making thing compact when I can but it's not a hard and fast rule.
Merging these changes.
